pylot/control/mpc/mpc_agent_operator.py

```diff
@@ -0,0 +1,333 @@
+            'stop_traffic_lights': speed_factor_tl
+        }
+
+        return speed_factor, state
+
+    def setup_mpc(self, waypoints):
+        path = np.array([[wp.location.x, wp.location.y] for wp in waypoints])
+
+        # convert target waypoints into spline
+        spline = CubicSpline2D(path[:, 0], path[:, 1], self._logger)
+        ss = []
+        vels = []
+        xs = []
+        ys = []
+        yaws = []
+        ks = []
+        for s in spline.s[:-2]:
+            x, y = spline.calc_position(s)
+            yaw = np.abs(spline.calc_yaw(s))
+            k = spline.calc_curvature(s)
+            xs.append(x)
+            ys.append(y)
+            yaws.append(yaw)
+            ks.append(k)
+            ss.append(s)
+            vels.append(18)
+
+        self._config["reference"] = {
+            't_list': [],  # Time [s]
+            's_list': ss,  # Arc distance [m]
+            'x_list': xs,  # Desired X coordinates [m]
+            'y_list': ys,  # Desired Y coordinates [m]
+            'k_list': ks,  # Curvatures [1/m]
+            'vel_list': vels,  # Desired tangential velocities [m/s]
+            'yaw_list': yaws,  # Yaws [rad]
+        }
+        # draw intended trajectory
+        for p in path:
+            self._world.debug.draw_point(carla.Location(x=p[0], y=p[1], z=0.5),
+                                         size=0.2,
+                                         life_time=1.0, color=carla.Color(0, 0, 255))
+
+        # initialize mpc controller
+        self.mpc = ModelPredictiveController(config=self._config)
+
+    def get_control_message(self, waypoints, vehicle_transform, current_speed, speed_factor, timestamp):
+        # Figure out the location of the ego vehicle and compute the next waypoint.
+        ego_location = to_carla_location(vehicle_transform.location)
+        ego_yaw = np.deg2rad(zero_to_2_pi(vehicle_transform.rotation.yaw))
+
+        # step the controller
+        self.setup_mpc(waypoints)
+        self.mpc.vehicle.x = ego_location.x
+        self.mpc.vehicle.y = ego_location.y
+        self.mpc.vehicle.yaw = ego_yaw
+
+        try:
+            self.mpc.step()
+        except Exception as e:
+            self._logger.info("Failed to solve MPC.")
+            self._logger.info(e)
+            return ControlMessage(0, 0, 1, False, False, timestamp)
+
+        # compute pid controls
+        target_speed = self.mpc.solution.vel_list[0] * speed_factor
+        target_yaw = self.mpc.solution.yaw_list[0]
+        target_steer_rad = self.mpc.horizon_steer[0]  # in rad
+        steer = self.__rad2steer(target_steer_rad)  # [-1.0, 1.0]
+        throttle, brake = self.__get_throttle_brake_without_factor(
+            current_speed, target_speed)
+
+        # draw next waypoints
+        self._world.debug.draw_point(carla.Location(x=ego_location.x, y=ego_location.y, z=0.5),
+                                     size=0.2,
+                                     life_time=30000.0)
```

Comment:
Same as above. We should look into defining a visualizer for MPC.

Reply:
Will defer to a later PR!